### PR TITLE
Device monitor

### PR DESCRIPTION
**Pull request by ferjm**
This is required for https://github.com/servo/servo/pull/27152/files

### CODE REVIEW

File: backends/gstreamer/device_monitor.rs [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+            .iter()
+            .map(|device| {
+                let display_name = device.get_display_name().as_str().to_owned();
+                MediaDeviceInfo {
+                    device_id: display_name.clone(),
+                    kind: match device.get_device_class().as_str() {
+                        AUDIO_SOURCE => MediaDeviceKind::AudioInput,
+                        AUDIO_SINK => MediaDeviceKind::AudioOutput,
+                        VIDEO_SOURCE => MediaDeviceKind::VideoInput,
+                        _ => MediaDeviceKind::__Unknown,
```
Comment (Manishearth) [inline backends/gstreamer/device_monitor.rs:42]: We should not have this variant, we should just filter_map these out

File: backends/gstreamer/device_monitor.rs [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+use gst::Caps;
+use gst::DeviceExt;
+use gst::DeviceMonitor as GstDeviceMonitor;
+use gst::DeviceMonitorExt;
+use gst::DeviceMonitorExtManual;
+
+use servo_media::{MediaDeviceInfo, MediaDeviceKind};
+
+pub struct GStreamerDeviceMonitor {
+    devices: AtomicRefCell<Option<Vec<MediaDeviceInfo>>>,
```
Comment (Manishearth) [inline backends/gstreamer/device_monitor.rs:11]: is the atomic refcell necessary?

can we instead return a GSTDeviceMonitor and have `navigator.mediadevices` contain one via a trait object?

Comment (ferjm) [reply backends/gstreamer/device_monitor.rs:11]: Sorry, I can't see how returning a `GstDeviceMonitor` makes the atomic refcell go away. The atomic refcell is here just to allow caching the result of the `GStreamerDeviceMonitor.enumerate_devices` call. If you think it is preferable to do the caching on the `navigator.mediaDevices` side, we can move the cache there, but I don't see the need to return the extra trait object.

Comment (Manishearth) [reply backends/gstreamer/device_monitor.rs:11]: Yeah I want to do the caching on the mediadevices side.

The trait object lets us return a DeviceMonitor object that contains the cache (and will only be accessed from one thread), but it's also fine if we do the caching in servo.

Might need tweaks once we decide to support devicechange.

File: backends/gstreamer/lib.rs [hunk 3]
```diff
@@ -1,4 +1,6 @@
 #![feature(nll)]
+
+extern crate atomic_refcell;
```
Comment (Manishearth) [inline backends/gstreamer/lib.rs:3]: unnecessary import

File: backends/gstreamer/device_monitor.rs [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+            })
+            .collect();
+        Ok(devices)
+    }
+}
+
+impl MediaDeviceMonitor for GStreamerDeviceMonitor {
+    fn enumerate_devices(&self) -> Result<Vec<MediaDeviceInfo>, ()> {
+        {
+            if let Some(ref devices) = *self.devices.borrow() {
```
Comment (Manishearth) [inline backends/gstreamer/device_monitor.rs:55]: we should probably file a followup about the devices changing

Comment (Manishearth) [reply backends/gstreamer/device_monitor.rs:55]: though it's not that important, actually, sicne we don't use this cache in servo

**Pull request merged**